### PR TITLE
Keytable filter

### PR DESCRIPTION
**Pull request by catoverdrive**
@cseed once #2573 I'll probably change this to use that stuff.

### CODE REVIEW

File: src/main/scala/is/hail/expr/Relational.scala [hunk 1]
```diff
@@ -565,3 +593,25 @@ case class KeyTableLiteral(value: KeyTableValue) extends KeyTableIR {
...
   def execute(hc: HailContext): KeyTableValue = value
 }
+
+case class FilterKT(child: KeyTableIR, pred: IR) extends KeyTableIR {
+  def children: IndexedSeq[BaseIR] = Array(child, pred)
+
+  def typ: KeyTableType = child.typ
+
+  def copy(newChildren: IndexedSeq[BaseIR]): FilterKT = {
+    assert(newChildren.length == 1)
```
Comment (cseed) [inline src/main/scala/is/hail/expr/Relational.scala]: Length should be two.  We need to rewrite the pred, too.

File: src/main/scala/is/hail/keytable/KeyTable.scala [hunk 2]
```diff
@@ -453,19 +454,45 @@ class KeyTable(val hc: HailContext,
...
   }
 
-  def filter(p: Annotation => Boolean): KeyTable = copy(rdd = rdd.filter(p))
-
   def filter(cond: String, keep: Boolean): KeyTable = {
-    val ec = rowEvalContext()
+    var filterAST = Parser.expr.parse(cond)
+    val pred = filterAST.toIR
+    pred match {
+      case None => {
```
Comment (tpoterba) [inline src/main/scala/is/hail/keytable/KeyTable.scala]: can you invert the Some and None cases here?

Comment (tpoterba) [reply src/main/scala/is/hail/keytable/KeyTable.scala]: I think it's a lot more conventional.

File: src/main/scala/is/hail/keytable/KeyTable.scala [hunk 3]
```diff
@@ -453,19 +454,45 @@ class KeyTable(val hc: HailContext,
...
+            i += 1
+          }
+          val ret = f()
+          ret != null && ret.booleanValue()
+        }
 
-    filter(p)
+        copy2(rvd = rvd.filter(p))
+      }
+      case Some(irPred) => {
```
Comment (cseed) [inline src/main/scala/is/hail/keytable/KeyTable.scala]: { is unnecessary.  case statements make a block.

File: src/main/scala/is/hail/keytable/KeyTable.scala [hunk 4]
```diff
@@ -453,19 +454,45 @@ class KeyTable(val hc: HailContext,
...
+        new KeyTable(hc, FilterKT(ir, if (keep) irPred else ApplyUnaryPrimOp(Bang(), irPred, TBoolean())))
```
Comment (cseed) [inline src/main/scala/is/hail/keytable/KeyTable.scala]: type is optional

File: src/test/scala/is/hail/methods/KeyTableSuite.scala [hunk 5]
```diff
@@ -588,6 +588,10 @@ class KeyTableSuite extends SparkSuite {
       .annotateGlobal(Map(5 -> "bar"), TDict(TInt32Optional, TStringOptional), "dict")
       .annotateGlobalExpr("another = foo[1]")
 
+    println(kt.globalSignature)
```
Comment (tpoterba) [inline src/test/scala/is/hail/methods/KeyTableSuite.scala]: stray prints

File: src/main/scala/is/hail/expr/Relational.scala [hunk 6]
```diff
@@ -539,9 +540,36 @@ case class KeyTableValue(typ: KeyTableType, localValue: KTLocalValue, rvd: RVD)
     val localRowType = typ.rowType
     rvd.rdd.map { rv => new UnsafeRow(localRowType, rv.region.copy(), rv.offset) }
   }
+
+  def filter(p: () => (RegionValue, RegionValue) => Boolean): KeyTableValue = {
```
Comment (cseed) [inline src/main/scala/is/hail/expr/Relational.scala]: Why does this take a thunk?

File: src/test/scala/is/hail/methods/KeyTableSuite.scala [hunk 7]
```diff
@@ -145,17 +145,17 @@ class KeyTableSuite extends SparkSuite {
     assert(ktLeftJoin.count == ktLeft.count &&
       ktLeftJoin.nKeys == nExpectedKeys &&
       ktLeftJoin.nColumns == nExpectedColumns &&
-      ktLeftJoin.filter { a =>
+      ktLeftJoin.copy(rdd = ktLeftJoin.rdd.filter { a =>
```
Comment (tpoterba) [inline src/test/scala/is/hail/methods/KeyTableSuite.scala]: why this change?

File: src/main/scala/is/hail/expr/Relational.scala [hunk 8]
```diff
@@ -539,9 +540,36 @@ case class KeyTableValue(typ: KeyTableType, localValue: KTLocalValue, rvd: RVD)
...
+    val globalType = typ.globalType
+    val globals = localValue.globals
+    copy(rvd = rvd.mapPartitions(typ.rowType) { it =>
+      val f = p()
+      val rv2 = RegionValue()
```
Comment (cseed) [inline src/main/scala/is/hail/expr/Relational.scala]: call this `globalRV` maybe?

File: src/main/scala/is/hail/expr/Relational.scala [hunk 9]
```diff
@@ -565,3 +593,25 @@ case class KeyTableLiteral(value: KeyTableValue) extends KeyTableIR {
...
+    FilterKT(newChildren(0).asInstanceOf[KeyTableIR], pred)
+  }
+  def execute(hc: HailContext): KeyTableValue = {
+    val ktv = child.execute(hc)
+    val mappedPred = typ.remapIR(pred)
+    Infer(mappedPred)
+    val fb = FunctionBuilder.functionBuilder[Region, Long, Boolean, Long, Boolean, Boolean]
+    Emit(mappedPred, fb)
+    val f = fb.result()
+    ktv.filter(() => (rv, rv2) => {
```
Comment (cseed) [inline src/main/scala/is/hail/expr/Relational.scala]: { also unnecessary.

File: src/main/scala/is/hail/keytable/KeyTable.scala (hunk 4 above)
Comment (catoverdrive) [reply src/main/scala/is/hail/keytable/KeyTable.scala]: You mean the TBoolean() on this line?

File: src/test/scala/is/hail/methods/KeyTableSuite.scala (hunk 7 above)
Comment (catoverdrive) [reply src/test/scala/is/hail/methods/KeyTableSuite.scala]: I got rid of KeyTable.filter(annotation => boolean) method since this was the only place it was used.
